File: rust/src/api/features.rs
// this file is @generated
#[allow(unused_imports)]
use crate::{error::Result, models::*, Configuration};

#[derive(Default)]
pub struct FeaturesListFeaturesOptions {
    /// Filter by feature status. Repeat the param to select multiple, omit to return all.
    pub statuses: Option<Vec<FeatureStatus>>,

    /// Filter by product. Omit to return features across all products.
    pub product_id: Option<ProductId>,

    /// Search by feature name.
    pub search: Option<String>,

    /// Page number (0-indexed)
    pub page: Option<i32>,

    /// Number of items per page
    pub per_page: Option<i32>,
}

pub struct Features<'a> {
    cfg: &'a Configuration,
}

impl<'a> Features<'a> {
    pub(super) fn new(cfg: &'a Configuration) -> Self {
        Self { cfg }
    }

    pub async fn list_features(
        &self,
        options: Option<FeaturesListFeaturesOptions>,
    ) -> Result<crate::models::FeatureListResponse> {
        let FeaturesListFeaturesOptions {
            statuses,
            product_id,
            search,
            page,
            per_page,
        } = options.unwrap_or_default();

        crate::request::Request::new(http1::Method::GET, "/api/v1/features")
            .with_optional_exploded_query_param("statuses", statuses)
            .with_optional_query_param("product_id", product_id)
            .with_optional_query_param("search", search)
            .with_optional_query_param("page", page)
            .with_optional_query_param("per_page", per_page)
            .execute(self.cfg)
            .await
    }

    pub async fn get_feature(&self, feature_id: String) -> Result<crate::models::Feature> {
        crate::request::Request::new(http1::Method::GET, "/api/v1/features/{feature_id}")
            .with_path_param("feature_id", feature_id)
            .execute(self.cfg)
            .await
    }
}

File: typescript/src/api/features.ts
// this file is @generated
import { type Feature, FeatureSerializer } from "../models/feature";
import {
  type FeatureListResponse,
  FeatureListResponseSerializer,
} from "../models/featureListResponse";
import type { FeatureStatus } from "../models/featureStatus";
import type { ProductId } from "../models/productId";
import { HttpMethod, MeteroidRequest, type MeteroidRequestContext } from "../request";

export interface FeaturesListFeaturesOptions {
  /** Filter by feature status. Repeat the param to select multiple, omit to return all. */
  statuses?: FeatureStatus[];
  /** Filter by product. Omit to return features across all products. */
  productId?: ProductId;
  /** Search by feature name. */
  search?: string;
  /** Page number (0-indexed) */
  page?: number;
  /** Number of items per page */
  perPage?: number;
}

export class Features {
  public constructor(private readonly requestCtx: MeteroidRequestContext) {}

  /**  */
  public listFeatures(
    options?: FeaturesListFeaturesOptions
  ): Promise<FeatureListResponse> {
    const request = new MeteroidRequest(HttpMethod.GET, "/api/v1/features");

    request.setExplodedQueryParam("statuses", options?.statuses);
    request.setQueryParam("product_id", options?.productId);
    request.setQueryParam("search", options?.search);
    request.setQueryParam("page", options?.page);
    request.setQueryParam("per_page", options?.perPage);
    return request.send(this.requestCtx, FeatureListResponseSerializer._fromJsonObject);
  }

  /**  */
  public getFeature(idOrCode: string): Promise<Feature> {
    const request = new MeteroidRequest(HttpMethod.GET, "/api/v1/features/{id_or_code}");

    request.setPathParam("id_or_code", idOrCode);
    return request.send(this.requestCtx, FeatureSerializer._fromJsonObject);
  }
}
